File: android/core/src/main/kotlin/app/fihaven/core/logic/SpendingInsights.kt
package app.fihaven.core.logic

import app.fihaven.core.model.SpendTransaction
import java.time.LocalDate

/// Period-over-period category deltas (Pro) — port of spendingInsights.js.
object SpendingInsights {
    private val defaultCats = listOf(
        "Groceries", "Dining", "Shopping", "Transport", "Entertainment", "Health", "Bills", "Other",
    )

    data class Row(
        val cat: String,
        val now: Double,
        val was: Double,
        val delta: Double,
        val pct: Int,
    )

    fun spentByCategory(transactions: List<SpendTransaction>, bounds: PeriodBounds): Map<String, Double> {
        val m = mutableMapOf<String, Double>()
        transactions.forEach { t ->
            if (!transactionInPeriod(t.date, bounds)) return@forEach
            val cat = t.category.ifBlank { "Other" }
            m[cat] = (m[cat] ?: 0.0) + t.amount
        }
        return m
    }

    /** Compare current vs previous period spending by category (largest swings first). */
    fun compute(
        transactions: List<SpendTransaction>,
        currentBounds: PeriodBounds,
        prevBounds: PeriodBounds,
    ): List<Row> {
        val cur = spentByCategory(transactions, currentBounds)
        val prev = spentByCategory(transactions, prevBounds)
        val cats = (defaultCats + cur.keys + prev.keys).toSet()
        return cats.mapNotNull { cat ->
            val now = cur[cat] ?: 0.0
            val was = prev[cat] ?: 0.0
            if (now <= 0 && was <= 0) return@mapNotNull null
            val delta = now - was
            val pct = when {
                was > 0 -> kotlin.math.round(delta / was * 100).toInt()
                now > 0 -> 100
                else -> 0
            }
            Row(cat, now, was, delta, pct)
        }.sortedByDescending { kotlin.math.abs(it.delta) }
    }

    private fun transactionInPeriod(date: String, bounds: PeriodBounds): Boolean {
        if (date.isBlank()) return false
        val d = runCatching { LocalDate.parse(date) }.getOrNull() ?: return false
        return !d.isBefore(bounds.start) && d.isBefore(bounds.end)
    }
}

File: ios/FiHavenCore/Sources/FiHavenCore/Logic/SpendingInsights.swift
import Foundation

/// Period-over-period spending category deltas — port of spendingInsights.js.
public enum SpendingInsights {
    public struct Row: Equatable, Sendable {
        public var cat: String
        public var now: Double
        public var was: Double
        public var delta: Double
        public var pct: Int
    }

    private static let categories = spendingCategories

    public static func spentByCategory(_ transactions: [SpendTransaction], bounds: PeriodBounds) -> [String: Double] {
        var m: [String: Double] = [:]
        for t in transactions {
            guard BudgetRules.transactionInPeriod(t.date, bounds: bounds) else { continue }
            let cat = t.category.isEmpty ? "Other" : t.category
            m[cat, default: 0] += t.amount
        }
        return m
    }

    /// Compare current vs previous period spending by category, largest swings first.
    public static func compute(
        _ transactions: [SpendTransaction],
        currentBounds: PeriodBounds,
        prevBounds: PeriodBounds
    ) -> [Row] {
        let cur = spentByCategory(transactions, bounds: currentBounds)
        let prev = spentByCategory(transactions, bounds: prevBounds)
        var cats = Set(categories)
        cats.formUnion(cur.keys)
        cats.formUnion(prev.keys)

        var rows: [Row] = []
        for cat in cats {
            let now = cur[cat] ?? 0
            let was = prev[cat] ?? 0
            if now <= 0 && was <= 0 { continue }
            let delta = now - was
            let pct: Int
            if was > 0 {
                pct = Int((delta / was * 100).rounded())
            } else {
                pct = now > 0 ? 100 : 0
            }
            rows.append(Row(cat: cat, now: now, was: was, delta: delta, pct: pct))
        }
        rows.sort { abs($0.delta) > abs($1.delta) }
        return rows
    }
}
